File: ios/Classes/PrebidView.swift
import Foundation
import Flutter
import UIKit
import PrebidMobile
import GoogleMobileAds
import AppTrackingTransparency
import OSLog
import os

///A class that is responsible for creating and adding banner and interstitial ads to the Flutter app's view, as well as pausing and resuming auction
class PrebidView: NSObject, FlutterPlatformView, GADBannerViewDelegate, GADFullScreenContentDelegate {
    private var _methodChannel: FlutterMethodChannel
    private var bannerLayout: UIView!

    private var gamBanner: GAMBannerView!
    private var bannerAdUnit: BannerAdUnit!
    private var interstitialAdUnit: InterstitialAdUnit!

    private var errorText: String
    let customLog = OSLog(subsystem: "setupad.prebid.plugin", category: "PrebidPluginLog")

    ///Setting channel method, configuring banner layout and checking tracking permission
    init(frame: CGRect,
         viewIdentifier viewId: Int64,
         arguments args: Any?,
         binaryMessenger messenger: FlutterBinaryMessenger) {
        _methodChannel = FlutterMethodChannel(name: "setupad.plugin.setupad_prebid_flutter/myChannel_\(viewId)", binaryMessenger: messenger)
        errorText="No errors"
        super.init()

        _methodChannel.setMethodCallHandler(onMethodCall)

        bannerLayout=UIView(frame: frame)

        if #available(iOS 14, *) {
            ATTrackingManager.requestTrackingAuthorization { status in
                if status == .authorized {
                    os_log("Permission to track granted", log: self.customLog, type: .info)
                } else {
                    os_log("Permission to track denied", log: self.customLog, type: .info)
                }
            }
        } else {
            os_log("Not available on iOS versions prior to 14", log: customLog, type: .info)
        }
    }

    ///Adding a view to the UI where banner will be added
    func view() -> UIView {
        return bannerLayout
    }

    ///Disposing view and destroying Prebid auction
    func dispose() {
        bannerLayout = nil
        _methodChannel.setMethodCallHandler(nil)
        onDestroy()
    }

    ///Checking which method was called and then calling the corresponding method
    func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
        switch(call.method){
        case "setParams":
            settingParameters(call:call, result:result)
        case "pauseAuction":
            onPause()
        case "resumeAuction":
            onResume()
        case "destroyAuction":
            onDestroy()
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    ///Getting the parameters user wrote and then validating them
    func settingParameters(call: FlutterMethodCall, result: FlutterResult){
        os_log("Setting ad parameters", log: customLog, type: .info)
        let argument = call.arguments as! Dictionary<String, Any>
        let adType = argument["adType"] as? String ?? ""
        let adUnitId = argument["adUnitId"] as? String ?? ""
        let configId = argument["configId"] as? String ?? ""
        let width = argument["width"] as? Int ?? 0
        let height = argument["height"] as? Int ?? 0
        let refreshInterval = argument["refreshInterval"] as? Int ?? 0

        ///These two conditional statements are used to check if the ad type is not empy and correctly written
        if(adType == ""){
            os_log("Ad type is empty", log: customLog, type: .error)
        }
        else if (adType.lowercased() != "banner" && adType.lowercased() != "interstitial"){
            os_log("Ad type not recognized, did you mean banner or interstitial?", log: customLog, type: .error)
        }
        else{
            ///This switch case is used to check ad unit ID and config ID are not empty.
            ///If at least one of them is, an error is displayed in console.
            switch (adUnitId, configId){
            case ("", let config) where !config.isEmpty: /// Ad unit ID is empty
                os_log("Ad unit ID is empty", log: customLog, type: .error)

            case (let adUnit, "") where !adUnit.isEmpty: /// Config ID is empty
                os_log("Config ID is empty", log: customLog, type: .error)

            case ("", ""): /// Ad unit ID and config ID are empty
                os_log("Ad unit ID and config ID are empty", log: customLog, type: .error)

            case (_, _) where refreshInterval < 30 && adType.lowercased() != "interstitial": /// Refresh interval is less than 30 seconds and the ad type is not banner
                os_log("Refresh interval should be at least 30 seconds", log: customLog, type: .error)
                
            case (_, _) where refreshInterval > 120 && adType.lowercased() != "interstitial": /// Refresh interval is less than 30 seconds and the ad type is not banner
                os_log("Refresh interval should be no more than 120 seconds", log: customLog, type: .error)

            default: ///If there is no problem, either createBanner or createInterstitial is called
                os_log("Parameters set successfully!", log: customLog, type: .info)
                if (adType == "banner"){
                    createBanner(adUnitId, configId, width, height, refreshInterval)
                }
                else {
                    createInterstitial(adUnitId, configId, width, height)
                    ///Banner layout, which was created at the beggining of this class in the init method, and which is sized as a 1x1, is being hidden from the view
                    DispatchQueue.main.async {
                        self.bannerLayout.isHidden = true
                    }
                }
            }
        }
    }

    ///Setting banner parameters and fetching demand
    func createBanner(_ AD_UNIT_ID: String, _ CONFIG_ID: String, _ width: Int, _ height: Int, _ refreshInterval: Int){
        bannerAdUnit = BannerAdUnit(configId: CONFIG_ID, size: CGSize(width: width, height: height))
        bannerAdUnit.setAutoRefreshMillis(time: Double(refreshInterval*1000))

        let parameters = BannerParameters()
        parameters.api = [Signals.Api.MRAID_3]
        bannerAdUnit.bannerParameters = parameters

        ///DispatchQueue is used to avoid problems when changing banner layout (by adding a gamBanner to it) due to the banner ayout being in main thread
        DispatchQueue.main.async { [self] in
            gamBanner = GAMBannerView(adSize: GADAdSizeFromCGSize(CGSize(width: width, height: height)))
            gamBanner.adUnitID=AD_UNIT_ID
            gamBanner.rootViewController = UIApplication.shared.delegate!.window!!.rootViewController!
            gamBanner.delegate = self
            addBannerViewToView(gamBanner)
            let request = GAMRequest()
            bannerAdUnit.fetchDemand(adObject: request){ [weak self] resultCode in
                os_log("Prebid demand fetch for GAM result: %@",
                       log: self!.customLog,
                       type: .info,
                       resultCode.name())
                self!.gamBanner.load(request)
            }
        }
    }

    ///Setting interstitial ad parameters and fetching demand
    func createInterstitial(_ AD_UNIT_ID: String, _ CONFIG_ID: String, _ width: Int, _ height: Int){
        if(width != 0 && height != 0){
            interstitialAdUnit = InterstitialAdUnit(configId: CONFIG_ID, minWidthPerc: width, minHeightPerc: height)
        }
        else{
            interstitialAdUnit = InterstitialAdUnit(configId: CONFIG_ID)
        }
            
        let gamRequest = GAMRequest()
        interstitialAdUnit.fetchDemand(adObject: gamRequest) { [weak self] resultCode in
            os_log("Prebid demand fetch for GAM result: %@",
                   log: self!.customLog,
                   type: .info,
                   resultCode.name())

            GAMInterstitialAd.load(withAdManagerAdUnitID: AD_UNIT_ID, request: gamRequest) { ad, error in
                guard let self = self else { return }

                if let error = error {
                    os_log("%{error}: Failed to load interstitial ad with error: %@",
                           log: self.customLog,
                           type: .error,
                           error.localizedDescription)
                } else if let ad = ad {
                    ad.fullScreenContentDelegate = self
                    guard let rootViewController = UIApplication.shared.delegate?.window??.rootViewController else {
                        os_log("Error: Could not find root view controller to present the ad", log: self.customLog, type: .error)
                        return
                    }
                    ad.present(fromRootViewController: rootViewController)
                }
            }
        }
    }

    ///Resizing banner
    func bannerViewDidReceiveAd(_ bannerView: GADBannerView) {
        AdViewUtils.findPrebidCreativeSize(bannerView, success: { size in
            guard let bannerView = bannerView as? GAMBannerView else { return }
            bannerView.resize(GADAdSizeFromCGSize(size))
            os_log("Success in finding Prebid creative size", log: self.customLog, type: .info)
        }, failure: { (error) in
            os_log("Failure in finding Prebid creative size: %{error} ",
                   log: self.customLog,
                   type: .error,
                   error.localizedDescription)
        })
    }

    ///Adding banner to the banner layout
    func addBannerViewToView(_ bannerView: GAMBannerView) {
        bannerLayout.translatesAutoresizingMaskIntoConstraints = false
        bannerLayout.addSubview(bannerView)
        bannerLayout.addConstraints(
            [NSLayoutConstraint(item: bannerView,
                                attribute: .bottom,
                                relatedBy: .equal,
                                toItem: bannerLayout.safeAreaLayoutGuide,
                                attribute: .bottom,
                                multiplier: 1,
                                constant: 0),
             NSLayoutConstraint(item: bannerView,
                                attribute: .centerX,
                                relatedBy: .equal,
                                toItem: bannerLayout,
                                attribute: .centerX,
                                multiplier: 1,
                                constant: 0)
            ])
    }

    ///A function for resuming auction
    func onResume(){
        os_log("Resuming Prebid auction", log: customLog, type: .info)
        bannerAdUnit?.resumeAutoRefresh()
    }

    ///A function for stopping auction
    func onPause(){
        os_log("Pausing Prebid auction", log: customLog, type: .info)
        bannerAdUnit?.stopAutoRefresh()
    }

    ///A function for stopping auction and removing reference to the bannerAdUnit, as well as hiding banner layout from the view
    func onDestroy() {
        bannerAdUnit?.stopAutoRefresh()
        os_log("Destroying Prebid auction", log: customLog, type: .info)
        bannerAdUnit = nil
    }
}

File: android/src/main/kotlin/setupad/plugin/setupad_prebid_flutter/PrebidView.kt
package setupad.plugin.setupad_prebid_flutter

import android.app.Activity
import android.content.Context
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.admanager.AdManagerAdRequest
import com.google.android.gms.ads.admanager.AdManagerAdView
import com.google.android.gms.ads.admanager.AdManagerInterstitialAd
import com.google.android.gms.ads.admanager.AdManagerInterstitialAdLoadCallback
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.platform.PlatformView
import org.prebid.mobile.BannerAdUnit
import org.prebid.mobile.BannerParameters
import org.prebid.mobile.InterstitialAdUnit
import org.prebid.mobile.Signals
import org.prebid.mobile.addendum.AdViewUtils
import org.prebid.mobile.addendum.PbFindSizeError

/**
 * A class that is responsible for creating and adding banner and interstitial ads to the Flutter app's view, as well as pausing and resuming auction
 */
class PrebidView internal constructor(
    context: Context,
    messenger: BinaryMessenger,
    id: Int,
    activity: Activity
) :
    PlatformView, MethodCallHandler {
    private var applicationContext: Context = context
    private var appActivity: Activity = activity
    private val channel = MethodChannel(messenger, "setupad.plugin.setupad_prebid_flutter/myChannel_$id")
    private var bannerLayout: ViewGroup?

    private var bannerAdUnit: BannerAdUnit? = null
    private var interstitialAdUnit: InterstitialAdUnit? = null

    private val Tag = "PrebidPluginLog"

    /**
     * Setting channel method, configuring banner layout
     */
    init {
        try {
            channel.setMethodCallHandler(this)
        } catch (e: Exception) {
            Log.e(Tag, "Error setting method call handler: $e")
        }
        bannerLayout = FrameLayout(applicationContext)
        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        )
        params.gravity = Gravity.BOTTOM
    }

    /**
     * Adding a view to the UI where banner will be added
     */
    override fun getView(): View? {
        return bannerLayout
    }

    /**
     * Disposing view and destroying Prebid auction
     */
    override fun dispose() {
        bannerLayout = null
        channel.setMethodCallHandler(null)
        onDestroy()
    }

    /**
     * Checking which method was called and then calling the needed method
     */
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setParams" -> settingParameters(call)
            "pauseAuction" -> onPause()
            "resumeAuction" -> onResume()
            "destroyAuction" -> onDestroy()
            else -> result.notImplemented()
        }
    }

    /**
     * Getting the parameters user wrote and then validating them
     */
    private fun settingParameters(call: MethodCall) {
        Log.d(Tag, "Setting ad parameters")
        val arguments = call.arguments as? Map<*, *>
        val adType = arguments?.get("adType") as? String ?: ""
        val adUnitId = arguments?.get("adUnitId") as? String ?: ""
        val configId = arguments?.get("configId") as? String ?: ""
        val width = arguments?.get("width") as? Int ?: 0
        val height = arguments?.get("height") as? Int ?: 0
        val refreshInterval = arguments?.get("refreshInterval") as? Int ?: 0

        bannerLayout?.removeAllViews()

        when {
            adType == "" -> {
                Log.e(Tag, applicationContext.getString(R.string.emptyAdType))
            }

            adType.lowercase() != "banner" && adType.lowercase() != "interstitial" -> {
                Log.e(Tag, applicationContext.getString(R.string.errorsInAdTypeName))
            }

            else -> {
                when {
                    adUnitId == "" && configId != "" -> { //adUnitID tuscias
                        Log.e(Tag, applicationContext.getString(R.string.emptyAdUnitID))
                    }

                    adUnitId != "" && configId == "" -> { //configID tuscias
                        Log.e(Tag, applicationContext.getString(R.string.emptyConfigID))
                    }

                    adUnitId == "" && configId == "" -> { //ad unit ir config ID tusti
                        Log.e(Tag, applicationContext.getString(R.string.emptyAdUnitConfigID))
                    }

                    refreshInterval < 30 && adType.lowercase()!="interstitial" -> {
                    Log.e(Tag, applicationContext.getString(R.string.tooSmallRefreshInterval))
                    }

                    refreshInterval > 120 && adType.lowercase()!="interstitial" -> {
                        Log.e(Tag, applicationContext.getString(R.string.tooBigRefreshInterval))
                    }

                    else -> {
                        Log.d(Tag, "Parameters set successfully!")
                        if (adType.lowercase() == "banner"){
                            createBanner(adUnitId, configId, width, height, refreshInterval)
                        }

                        else {
                            createInterstitial(adUnitId, configId, width, height)
                            bannerLayout?.visibility = View.GONE
                        }
                    }
                }
            }
        }
    }

    /**
     * Setting banner parameters and fetching demand
     */
    private fun createBanner(
        AD_UNIT_ID: String,
        CONFIG_ID: String,
        width: Int,
        height: Int,
        refreshInterval: Int
    ) {
        bannerAdUnit = BannerAdUnit(CONFIG_ID, width, height)
        bannerAdUnit?.setAutoRefreshInterval(refreshInterval)

        val bannerParameters = BannerParameters()
        bannerParameters.api = listOf(Signals.Api.MRAID_3, Signals.Api.OMID_1)
        bannerAdUnit!!.bannerParameters = bannerParameters

        val bannerAdView = AdManagerAdView(applicationContext)
        bannerAdView.adUnitId = AD_UNIT_ID
        bannerAdView.setAdSizes(AdSize(width, height))
        bannerAdView.adListener = bannerListener(bannerAdView)

        bannerLayout?.addView(bannerAdView)

        val request = AdManagerAdRequest.Builder().build()
        bannerAdUnit?.fetchDemand(request) { result ->
            Log.d(Tag, "Prebid demand fetch for GAM result: $result")
            bannerAdView.loadAd(request)
        }
    }

    /**
     * Setting interstitial ad parameters and fetching demand
     */
    private fun createInterstitial(AD_UNIT_ID: String, CONFIG_ID: String, width: Int, height: Int) {
        if (width!=0 && height!=0){
            interstitialAdUnit = InterstitialAdUnit(
                CONFIG_ID,
                width,
                height
            )
        }
        else{
            interstitialAdUnit = InterstitialAdUnit(CONFIG_ID)
        }


        val request = AdManagerAdRequest.Builder().build()
        interstitialAdUnit?.fetchDemand(request) { result ->
            Log.d(Tag, "Prebid demand fetch for GAM result: $result")
            AdManagerInterstitialAd.load(
                applicationContext,
                AD_UNIT_ID,
                request,
                interstitialListener()
            )
        }
    }

    /**
     * Banner listener which, if ad is loaded, resizes that banner
     * findPrebidCreativeSize() method is a fix for GAM bug, where the ad is sized incorrectly
     */
    private fun bannerListener(bannerAdView: AdManagerAdView): AdListener {
        return object : AdListener() {
            override fun onAdLoaded() {
                super.onAdLoaded()
                AdViewUtils.findPrebidCreativeSize(
                    bannerAdView,
                    object : AdViewUtils.PbFindSizeListener {
                        override fun success(width: Int, height: Int) {
                            Log.d(
                                Tag,
                                "Success in finding Prebid creative size"
                            )
                            bannerAdView.setAdSizes(AdSize(width, height))
                        }

                        override fun failure(error: PbFindSizeError) {
                            Log.e(Tag, "Failure in finding Prebid creative size: $error")
                        }
                    })
            }

            override fun onAdFailedToLoad(error: LoadAdError) {
                super.onAdFailedToLoad(error)
                Log.e(Tag, "Failure in loading banner ad: $error")
            }
        }
    }

    /**
     * Interstitial ad listener that, if ad is loaded, shows that ad
     */
    private fun interstitialListener(): AdManagerInterstitialAdLoadCallback {
        return object : AdManagerInterstitialAdLoadCallback() {
            override fun onAdLoaded(adManagerInterstitialAd: AdManagerInterstitialAd) {
                super.onAdLoaded(adManagerInterstitialAd)
                adManagerInterstitialAd.show(appActivity)
            }

            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                super.onAdFailedToLoad(loadAdError)
                Log.e(Tag, "Ad failed to load: $loadAdError")
            }
        }
    }

    /**
     * A method for stopping auction and removing reference to the bannerAdUnit, as well as hiding
     * banner from the layout
     */
    private fun onDestroy() {
        if (bannerAdUnit != null) {
            bannerAdUnit!!.stopAutoRefresh()
            Log.d(Tag, "Destroying Prebid auction")
            bannerAdUnit = null
        }
    }

    /**
     * A method for pausing auction
     */
    private fun onPause() {
        if (bannerAdUnit != null) {
            bannerAdUnit!!.stopAutoRefresh()
            Log.d(Tag, "Pausing Prebid auction")
        }
    }

    /**
     * A method for resuming auction
     */
    private fun onResume() {
        if (bannerAdUnit != null) {
            Log.d(Tag, "Resuming Prebid auction")
            bannerAdUnit!!.resumeAutoRefresh()
        }
    }
}